### PR TITLE
Add cmake support.

### PR DESCRIPTION
**Pull request by xfxyjwf**
Add cmake support primarily for Visual Studio users. These scripts can also be used to build protobuf on Linux, MinGW, Cygwin as well.

This PR also includes a shell script used to copy file lists from src/Makefile.am to cmake files so we don't need to do that manually.

@pherl to review

### CODE REVIEW

File: cmake/README.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+   contains gtest already and copying the gest directory from there to your
+   protobuf directory:
+       https://github.com/google/protobuf/releases
+1) Use cmake to generate MSVC project files. Running the following commands
+   in a command shell will generate project files for Visual Studio 2008 in
+   a sub-directory named "build".
+     > cd path/to/protobuf/cmake
+     > mkdir build
+     > cd build
+     > cmake -G "Visual Studio 9 2008" ..
```
Comment (pherl) [inline cmake/README.md]: It looks like text after '..' becomes grey for some reason. Please double check if it is some special mark down syntax.

**Comment by pherl** [comment]
Should we delete the current msvc proj files? Or should we generated them in releases and do not include them in git?

**Comment by paleozogt** [comment]
I vote for leaving them out of git.  The generated project files can be thought of in the same way as binaries-- they're compiled artifacts generated from the source.

**Comment by jtattermusch** [comment]
I think commiting the generated projects makes sense. It's easier for users to start when they already have their vsprojects ready once they clone - non everyone has cmake installed.

if we commit the vsproject files, we need to make sure they are always up to date (there should be a presubmit check that verifies that regenerating the projects using cmake yields no diff).

File: cmake/README.md (hunk 1 above)
Comment (xfxyjwf) [inline cmake/README.md]: Done. Fixed the formatting of this file.

**Comment by xfxyjwf** [comment]
Vsproject files generated by cmake actually depend on cmake. When you build the solution, it will run cmake to check whether any dependencies have changed and regenerate vsproject files if necessary. The last time I checked, they have no plan to update cmake to generate *clean" vsproject files that can work alone without cmake. So users would have to install cmake anyway regardless of whether we check-in the generated files. Given this, checking in these files provides little value and we probably shouldn't do that.

**Comment by pherl** [comment]
LGTM

**Pull request merged**